### PR TITLE
sql: test statement timeout cancelation results in admission q removal

### LINKED ISSUE #73570: sql: verify statement timeout gets propagated to all admission control queues
This issue tracks adding tests that verify that once a statement times out, operations that might have been waiting in an admission control queue are canceled.

Currently the SQL layer cancels the statement when the timeout expires, but it's not clear to me if that cancelation is propagated to KV operations that are already queued up.

Epic: CRDB-10304

Jira issue: CRDB-11650

### PR DESCRIPTION
**Pull request by cucaroach**
Make sure that if a request is stuck in the admission queue and the
statement that created that request is canceled that the cancelation
propagates across the GRPC layer.

Also update contexts.md to elaborate on how this happens.

Fixes: #73570

Release justification: new test w/ very small non-test changes.

Release note: None

### CODE REVIEW

**Review by cucaroach — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @sumeerbhola)

___
*[pkg/server/node.go, line 1028 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mw9QUVk8TvegtfjvI3s:-Mw9QUVk8TvegtfjvI3t:bgh3q91) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce9a3436ea3cc1c9ad7186ac707c0e1292d18853/pkg/server/node.go#L1028)):*
> ```Go
> 	}
> 	br, err := n.batchInternal(ctx, tenantID, args)
> 	n.admissionController.AdmittedKVWorkDone(handle)
> ```

I'm not entirely sure if this error handling makes sense but it does dove tail nicely with a follow on change I was considering to push the admission control stuff down into batchInternal, the reason to do that is the admission control will run under the context of the tracing ctx and any admission control tracing would be recorded (currently its not).

File: pkg/kv/kvserver/testing_knobs.go
Comment (tbg) [inline pkg/kv/kvserver/testing_knobs.go:100]: Can you move this next to here:

https://github.com/cockroachdb/cockroach/blob/7051448458ce86c4ed8d9224921ae4ffd65bdac5/pkg/kv/kvserver/testing_knobs.go#L80-L83

and also name it symmetrically, i.e. something like `TestingResponseErrFilter`?

Comment (tbg) [reply pkg/kv/kvserver/testing_knobs.go:100]: Sounds good thanks!

**Review by msirek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @RaduBerinde, and @sumeerbhola)

___
*[pkg/server/node.go, line 1051 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwZ9kF1BO6e9_SVQTkz:-MwZ9kF1BO6e9_SVQTl-:b-cgcdfr) ([raw file](https://github.com/cockroachdb/cockroach/blob/eead1434ec3ec2b4ee6881b28e90455bb785e3b3/pkg/server/node.go#L1051)):*
> ```Go
> 		n.testingErrorEvent(ctx, args, errors.DecodeError(ctx, br.Error.EncodedError))
> 	}
> 	return br, nil
> ```

Is this code trying to call `testingErrorEvent` only for cases where `br` is set within the immediately-preceding `if` block?  That block might cover up any previous errors occurring during the call to `batchInternal` due to the line `br.Error = roachpb.NewError(err)`.  Not sure if this is intentional or not.  Maybe you want 2 calls to `testingErrorEvent`?



_Code quote:_
```Go
	if buildutil.CrdbTestBuild && br.Error != nil && n.testingErrorEvent != nil {
		n.testingErrorEvent(ctx, args, errors.DecodeError(ctx, br.Error.EncodedError))
	}
```
___
*[pkg/server/server.go, line 356 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwZC5FEBBa1KH1dfdKX:-MwZC5FEBBa1KH1dfdKY:b4z70tg) ([raw file](https://github.com/cockroachdb/cockroach/blob/eead1434ec3ec2b4ee6881b28e90455bb785e3b3/pkg/server/server.go#L356)):*
> ```Go
> 	tcsFactory := kvcoord.NewTxnCoordSenderFactory(txnCoordSenderFactoryCfg, distSender)
> 
> 	admissionOptions := admission.Merge(admission.DefaultOptions, cfg.TestingKnobs.AdmissionControl.(*admission.Options))
> ```

Is `cfg.TestingKnobs.AdmissionControl` guaranteed to always be non-nil here?  Would a nil-pointer check be useful?
___
*[pkg/sql/run_control_test.go, line 755 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwYtScO0luzUvE7e7j4:-MwYtScPDSLmiynzZ782:bqls3k5) ([raw file](https://github.com/cockroachdb/cockroach/blob/eead1434ec3ec2b4ee6881b28e90455bb785e3b3/pkg/sql/run_control_test.go#L755)):*
> ```Go
> 
> 	var wg sync.WaitGroup
> 	// we want to wait for all the queue blockers to come back and the
> ```

nit: comment looks incomplete

**Review by cucaroach — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @msirek, @RaduBerinde, and @sumeerbhola)

___
*[pkg/server/node.go, line 1051 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwZ9kF1BO6e9_SVQTkz:-MwZZ-cFCTUqzudhGPld:bl0sqrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/eead1434ec3ec2b4ee6881b28e90455bb785e3b3/pkg/server/node.go#L1051)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Is this code trying to call `testingErrorEvent` only for cases where `br` is set within the immediately-preceding `if` block?  That block might cover up any previous errors occurring during the call to `batchInternal` due to the line `br.Error = roachpb.NewError(err)`.  Not sure if this is intentional or not.  Maybe you want 2 calls to `testingErrorEvent`?
</blockquote></details>

If err is non-nil and br.Error is too we'll do a log.Fatalf, so its either one or the other right?  log.Fatalf will exit the process.
___
*[pkg/server/server.go, line 356 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwZC5FEBBa1KH1dfdKX:-MwZmJOLAQchGgVC6Y9q:b-bh5lkl) ([raw file](https://github.com/cockroachdb/cockroach/blob/eead1434ec3ec2b4ee6881b28e90455bb785e3b3/pkg/server/server.go#L356)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Is `cfg.TestingKnobs.AdmissionControl` guaranteed to always be non-nil here?  Would a nil-pointer check be useful?
</blockquote></details>

Only in a test environment, nice catch!   I don't know how I missed this, need to look into team city failures more apparently...
___
*[pkg/sql/run_control_test.go, line 755 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwYtScO0luzUvE7e7j4:-MwZmlJICCZ60pS2UPRI:b-8fx2ut) ([raw file](https://github.com/cockroachdb/cockroach/blob/eead1434ec3ec2b4ee6881b28e90455bb785e3b3/pkg/sql/run_control_test.go#L755)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

nit: comment looks incomplete
</blockquote></details>

:+1:

**Review by msirek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @msirek, @RaduBerinde, and @sumeerbhola)

___
*[pkg/server/node.go, line 1051 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwZ9kF1BO6e9_SVQTkz:-Mw_2X-B-Kl4Uhury5o3:b-fg6sp) ([raw file](https://github.com/cockroachdb/cockroach/blob/eead1434ec3ec2b4ee6881b28e90455bb785e3b3/pkg/server/node.go#L1051)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

If err is non-nil and br.Error is too we'll do a log.Fatalf, so its either one or the other right?  log.Fatalf will exit the process.
</blockquote></details>

Right.  I guess I'm asking, in that situation, do you want `testingErrorEvent` to get called  before `log.Fatalf` is called (in case `testingErrorEvent` captures any information that would be useful, which`log.Fatalf` doesn't capture)?

**Review by cucaroach — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @msirek, @RaduBerinde, and @sumeerbhola)

___
*[pkg/server/node.go, line 1051 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwZ9kF1BO6e9_SVQTkz:-Mwanlwi5Qc7MLLHbN9b:bq4gunl) ([raw file](https://github.com/cockroachdb/cockroach/blob/eead1434ec3ec2b4ee6881b28e90455bb785e3b3/pkg/server/node.go#L1051)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Right.  I guess I'm asking, in that situation, do you want `testingErrorEvent` to get called  before `log.Fatalf` is called (in case `testingErrorEvent` captures any information that would be useful, which`log.Fatalf` doesn't capture)?
</blockquote></details>

Ah!  Yeah in that event something is super wrong and I don't think it matters if testing hooks are fired.  I guess I would say if something if super wrong its best to react ASAP rather than doing some other work then reacting.

**Review by rytaft — COMMENTED** [review]
Cool! I'll let someone who is more familiar with this code give approval, I just left some comments (mostly nits) below.

> Reviewed 4 of 10 files at r1, 2 of 3 files at r2, 2 of 3 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-:-MwaxgZJ7jBiJGdLtsqB:b-qkz00n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @msirek, @RaduBerinde, and @sumeerbhola)

___
*[docs/tech-notes/contexts.md, line 281 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mwaq8yK2MXQI0g7sStG:-Mwaq8yK2MXQI0g7sStH:b-3jx16j) ([raw file](https://github.com/cockroachdb/cockroach/blob/071683ac7f8687a31afad57ecd83e911bb671005/docs/tech-notes/contexts.md#L281)):*
> ```Markdown
>   goroutine spawn under you for propagating the parents cancellation,
>   so it can be expensive.
> - GRPC propagates cancel and deadline contexts across the wire.  Deadline
> ```

nit: use a single space after a period (here and below)
___
*[docs/tech-notes/contexts.md, line 286 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mwarhv9-eihzB-D1J3I:-Mwarhv9-eihzB-D1J3J:brywlic) ([raw file](https://github.com/cockroachdb/cockroach/blob/071683ac7f8687a31afad57ecd83e911bb671005/docs/tech-notes/contexts.md#L286)):*
> ```Markdown
>   sending a message GRPC does a [select](https://github.com/cockroachdb/vendored/blob/ed53f102e86b51c228b61d3ad54f1428982fe75e/google.golang.org/grpc/internal/transport/transport.go#L178) on the response and also listens
>   on the Done channel for the context passed it by the client. This in turn will
>   close the GRPC stream which will send an RST message the server which will
> ```

nit: RST message the server -> RST message to the server

Also - what's an RST message?
___
*[docs/tech-notes/contexts.md, line 288 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwarrAf5qKAGX3DTvR3:-MwarrAf5qKAGX3DTvR4:b-8nvggq) ([raw file](https://github.com/cockroachdb/cockroach/blob/071683ac7f8687a31afad57ecd83e911bb671005/docs/tech-notes/contexts.md#L288)):*
> ```Markdown
>   close the GRPC stream which will send an RST message the server which will
>   cause the server side to be [canceled](https://github.com/cockroachdb/vendored/blob/ed53f102e86b51c228b61d3ad54f1428982fe75e/google.golang.org/grpc/internal/transport/http2_server.go#L1231).
>   See also TestGRPCDeadlinePropagation and TestTenantStatementTimeoutAdmissionQueueCancelation for test that rely on this
> ```

nit: test -> tests
___
*[pkg/server/node.go, line 1051 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mwasjci7umzZPpSBd5F:-Mwasjci7umzZPpSBd5G:beyk02w) ([raw file](https://github.com/cockroachdb/cockroach/blob/071683ac7f8687a31afad57ecd83e911bb671005/pkg/server/node.go#L1051)):*
> ```Go
> 		n.testingErrorEvent(ctx, args, errors.DecodeError(ctx, br.Error.EncodedError))
> 	}
> 	return br, nil
> ```

Is it a problem that we're no longer returning `err` in case `AdmitKVWork` failed?
___
*[pkg/util/admission/granter.go, line 594 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwavAft8aNt0z6o7Wmu:-MwavAft8aNt0z6o7Wmv:bte7xpu) ([raw file](https://github.com/cockroachdb/cockroach/blob/071683ac7f8687a31afad57ecd83e911bb671005/pkg/util/admission/granter.go#L594)):*
> ```Go
> 	useGrantChains bool
> 
> 	// testing only
> ```

Can you expand this comment to explain what it does?
___
*[pkg/util/admission/granter.go, line 644 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mwatv_J8FnxHdGGyxml:-Mwatv_J8FnxHdGGyxmm:b-8s9cza) ([raw file](https://github.com/cockroachdb/cockroach/blob/071683ac7f8687a31afad57ecd83e911bb671005/pkg/util/admission/granter.go#L644)):*
> ```Go
> // Merge combines default options with override to update values that aren't
> // the Go default values.
> func Merge(defaults Options, override *Options) Options {
> ```

nit: Merge sounds like you'd be combining these in an "equitable" manner, e.g. by taking the max value between the defaults and the override. But that's not really what's happening here. I think it might be clearer if you make this a method on the `Options` struct, and call it `Override`. e.g.: 
```
func (o *Options) Override(other *Options) {
  ...
}
```
___
*[pkg/util/admission/granter.go, line 945 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwavXxhA9T31-_1qLOK:-MwavXxhA9T31-_1qLOL:b-iuy0gc) ([raw file](https://github.com/cockroachdb/cockroach/blob/071683ac7f8687a31afad57ecd83e911bb671005/pkg/util/admission/granter.go#L945)):*
> ```Go
> 	coord.granters[SQLSQLResponseWork].(*tokenGranter).refillBurstTokens(skipEnforcement)
> 	if coord.granters[KVWork] != nil {
> 		kvg := coord.granters[KVWork].(*kvGranter)
> ```

nit: you don't need this variable if `coord.disableSkipEnforcement` is true
___
*[pkg/sql/run_control_test.go, line 734 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwaxTAX88Urw9dy_YO6:-MwaxTAX88Urw9dy_YO7:ba081gt) ([raw file](https://github.com/cockroachdb/cockroach/blob/071683ac7f8687a31afad57ecd83e911bb671005/pkg/sql/run_control_test.go#L734)):*
> ```Go
> 	return db
> }
> func TestTenantStatementTimeoutAdmissionQueueCancelation(t *testing.T) {
> ```

I think this test could use a high level comment somewhere to explain what it's doing and why the test works
___
*[pkg/sql/run_control_test.go, line 773 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwawZYIBTBWMQmL1a49:-MwawZYIBTBWMQmL1a4A:bsqtwe6) ([raw file](https://github.com/cockroachdb/cockroach/blob/071683ac7f8687a31afad57ecd83e911bb671005/pkg/sql/run_control_test.go#L773)):*
> ```Go
> 			AdmissionControl: &admission.Options{
> 				MaxCPUSlots: numBlockers,
> 				// During testing if CPU isn't responsive and skipEnforcement turns off admission control queuing behavior,
> ```

nit: comments should be <= 80 characters wide
___
*[pkg/sql/run_control_test.go, line 832 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mwax6k5EIgH9KUGIX1Y:-Mwax6k5EIgH9KUGIX1Z:b-7uoonp) ([raw file](https://github.com/cockroachdb/cockroach/blob/071683ac7f8687a31afad57ecd83e911bb671005/pkg/sql/run_control_test.go#L832)):*
> ```Go
> 		}(r)
> 	}
> 	// wait till queue is full
> ```

nit: start comment with capital letter and end with a period

**Review by msirek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @msirek, @RaduBerinde, and @sumeerbhola)

___
*[pkg/server/node.go, line 1051 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwZ9kF1BO6e9_SVQTkz:-Mwb29O_5bqELMJ9zxo2:bkihy1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/eead1434ec3ec2b4ee6881b28e90455bb785e3b3/pkg/server/node.go#L1051)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Ah!  Yeah in that event something is super wrong and I don't think it matters if testing hooks are fired.  I guess I would say if something if super wrong its best to react ASAP rather than doing some other work then reacting.
</blockquote></details>

OK, just checking.  Thanks.

**Comment by rytaft** [comment]
___
*[pkg/sql/run_control_test.go, line 755 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwYtScO0luzUvE7e7j4:-Mwb3WaK7RAdpYIgCy_6:b-aaphbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/eead1434ec3ec2b4ee6881b28e90455bb785e3b3/pkg/sql/run_control_test.go#L755)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

:+1:
</blockquote></details>

I still don't really understand this sentence -- feels like there are some words out of place

**Review by cucaroach — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @msirek, @RaduBerinde, @rytaft, and @sumeerbhola)

___
*[pkg/util/admission/granter.go, line 594 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwavAft8aNt0z6o7Wmu:-MwdwsAK7gl9IR6LlKzk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/071683ac7f8687a31afad57ecd83e911bb671005/pkg/util/admission/granter.go#L594)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Can you expand this comment to explain what it does?
</blockquote></details>

Done.
___
*[pkg/sql/run_control_test.go, line 734 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-MwaxTAX88Urw9dy_YO6:-Mwb2m-EDMCw-rnxdWeo:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/071683ac7f8687a31afad57ecd83e911bb671005/pkg/sql/run_control_test.go#L734)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think this test could use a high level comment somewhere to explain what it's doing and why the test works
</blockquote></details>

:+1:

**Review by cucaroach — COMMENTED** [review]
Thanks for the great feedback, I think I addressed everything!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-:-MwdxjDt3mKzf06GKE82:b-hovwlb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @msirek, @RaduBerinde, @rytaft, and @sumeerbhola)

**Review by rytaft — COMMENTED** [review]
Thanks!

> Reviewed 3 of 4 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-:-MwfkLHg6Zfu2eXOfTID:b-ucfknh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @msirek, @RaduBerinde, and @sumeerbhola)

**Review by msirek — COMMENTED** [review]
The code looks good, but the build/test time has more than doubled (usually it takes around 50-55 minutes), so you may wish to do some profiling and see what's going on.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-:-Mwhu9_V0ft2HZHw3_kc:b-6dt2cs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @msirek, @RaduBerinde, @rytaft, and @sumeerbhola)

**Comment by cucaroach** [comment]
Switch @RaduBerinde to @yuzefovich .   @sumeerbhola could you review this too please?    Thanks!

**Review by yuzefovich — APPROVED** [review]
Nice! :lgtm: it'd be good to get an approval from Sumeer too though.

> Reviewed 3 of 10 files at r1, 2 of 3 files at r2, 1 of 3 files at r3, 2 of 4 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-:-Mx0MCIQ-4YnzqUynjO_:butwt86)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach, @msirek, @rytaft, and @sumeerbhola)

___
*[pkg/server/node.go, line 1028 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mw9QUVk8TvegtfjvI3s:-Mx0K_T_1w7Q1RuwutUo:b-rh3sp4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce9a3436ea3cc1c9ad7186ac707c0e1292d18853/pkg/server/node.go#L1028)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

I'm not entirely sure if this error handling makes sense but it does dove tail nicely with a follow on change I was considering to push the admission control stuff down into batchInternal, the reason to do that is the admission control will run under the context of the tracing ctx and any admission control tracing would be recorded (currently its not).
</blockquote></details>

Is this comment still relevant?
___
*[pkg/server/node.go, line 1051 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mwasjci7umzZPpSBd5F:-Mx0KdXR92l7h0S2f97f:bk87fc) ([raw file](https://github.com/cockroachdb/cockroach/blob/071683ac7f8687a31afad57ecd83e911bb671005/pkg/server/node.go#L1051)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is it a problem that we're no longer returning `err` in case `AdmitKVWork` failed?
</blockquote></details>

`err` will still be returned in `br.Error`, I'm guessing it's ok.
___
*[pkg/sql/run_control_test.go, line 754 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mx0Lk4X4qpdZZmnfRmp:-Mx0Lk4X4qpdZZmnfRmq:b-xtjlvj) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbc99d7be71af1326498e44caa1935c47b1902d/pkg/sql/run_control_test.go#L754)):*
> ```Go
> 	numBlockers := 4
> 
> 	// We can't get the tableID programmatically here.
> ```

nit: I'd also add that we do assert that the actual table ID matches this value below, after the server is started.

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> Reviewed 3 of 10 files at r1, 2 of 3 files at r2, 1 of 3 files at r3, 1 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @cucaroach, @msirek, and @rytaft)

___
*[pkg/server/node.go, line 1028 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mw9QUVk8TvegtfjvI3s:-Mx0LDU4BHepRnYPcBgE:b-lo2hw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce9a3436ea3cc1c9ad7186ac707c0e1292d18853/pkg/server/node.go#L1028)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Is this comment still relevant?
</blockquote></details>

I don't understand how CockroachDB's tracing instrumentation works, so just trying to understand what you are saying here.
- Which tracing ctx are you referring to in `batchInternal`?
- Do the current `log.Eventf` calls in work_queue.go not show up in traces?

Regarding this specific change, someone should KV should comment on whether returning a nil BatchResponse with an explicit error is different from returning an embedded error in a BatchResponse with a nil error.
I wonder whether the first means that it is an error that should be recognized as such by grpc, while the latter hides that error from grpc. The difference would matter if/when we have informational pages regarding error rates etc. that rely only on what grpc sees.
And I don't know if this affects the /tracez work by @andreimatei 
___
*[pkg/util/admission/granter.go, line 596 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mx0OVes1gbTpkVfek_M:-Mx0OVet1K2Asf-pj-oC:b-nn0kjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbc99d7be71af1326498e44caa1935c47b1902d/pkg/util/admission/granter.go#L596)):*
> ```Go
> 	// With unreliable clocks we will abandon slot enforcement, for testing
> 	// queuing behavior we need reliable slot enforcement so this testing hook
> 	// can be used to disable it.
> ```

Should there be a period before "for testing ..." -- I am having difficulty parsing?

This should say something like:
// The admission control code needs high sampling frequency of the cpu load, and turns off admission control enforcement when the sampling frequency is too low. For testing queueing behavior, we do not want the enforcement to be turned off in a non-deterministic manner, ...  
___
*[pkg/util/admission/granter.go, line 623 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mx0QcTgAm-PIgyqKyv8:-Mx0QcTgAm-PIgyqKyv9:b-350oxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbc99d7be71af1326498e44caa1935c47b1902d/pkg/util/admission/granter.go#L623)):*
> ```Go
> 	SQLStatementLeafStartWorkSlots int
> 	SQLStatementRootStartWorkSlots int
> 	DisableSkipEnforcement         bool
> ```

can you rename this to `TestingDisableSkipEnforcement` and also the same name change in the `GrantCoordinator` field.
___
*[pkg/util/admission/granter.go, line 639 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mx0Oq_F-o-ZM6rwwN94:-Mx0Oq_F-o-ZM6rwwN95:bpse6af) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbc99d7be71af1326498e44caa1935c47b1902d/pkg/util/admission/granter.go#L639)):*
> ```Go
> 	MaxCPUSlots:                    100000, /* TODO(sumeer): add cluster setting */
> 	SQLKVResponseBurstTokens:       100000, /* TODO(sumeer): add cluster setting */
> 	SQLSQLResponseBurstTokens:      100000, /* arbitrary, and unused */
> ```

can you change this `SQLSQLResponseBurstTokens` comment to the same TODO as above. It is not unused.
___
*[pkg/util/admission/granter.go, line 645 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mx0PDieCyaJgc45jPHr:-Mx0PDieCyaJgc45jPHs:beod1r5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbc99d7be71af1326498e44caa1935c47b1902d/pkg/util/admission/granter.go#L645)):*
> ```Go
> 
> // Override applies values from "override" to the receiver that differ from Go
> // defaults.
> ```

Is differing from Go defaults how we normally handle such overrides? What if one wanted to override something to 0?

**Review by cucaroach — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @cucaroach, @msirek, @sumeerbhola, and @yuzefovich)

___
*[pkg/server/node.go, line 1028 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mw9QUVk8TvegtfjvI3s:-Mx19VbX1gMZkMQNy8AF:b-gaanlt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce9a3436ea3cc1c9ad7186ac707c0e1292d18853/pkg/server/node.go#L1028)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't understand how CockroachDB's tracing instrumentation works, so just trying to understand what you are saying here.
- Which tracing ctx are you referring to in `batchInternal`?
- Do the current `log.Eventf` calls in work_queue.go not show up in traces?

Regarding this specific change, someone should KV should comment on whether returning a nil BatchResponse with an explicit error is different from returning an embedded error in a BatchResponse with a nil error.
I wonder whether the first means that it is an error that should be recognized as such by grpc, while the latter hides that error from grpc. The difference would matter if/when we have informational pages regarding error rates etc. that rely only on what grpc sees.
And I don't know if this affects the /tracez work by @andreimatei 

</blockquote></details>

To get the log.Eventf calls to "work" I had to push the admission control stuff under the context created in batchInternal so yeah I think as far as distributed tracing is concerned those log calls may be ignored.   Its immaterial in a cancelation context though as when the client side hangs up with a close (which is what happens when client side cancels) none of the response ever gets back to the client.
___
*[pkg/server/node.go, line 1051 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mwasjci7umzZPpSBd5F:-Mx1A6udCb_cGcthGUcm:bpyux5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/071683ac7f8687a31afad57ecd83e911bb671005/pkg/server/node.go#L1051)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

`err` will still be returned in `br.Error`, I'm guessing it's ok.
</blockquote></details>

Like @sumeerbhola said above it would be good for a kv person to sign off on my changes to the error handling.
___
*[pkg/util/admission/granter.go, line 623 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mx0QcTgAm-PIgyqKyv8:-Mx1AmAs1oavcLEaFAi8:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbc99d7be71af1326498e44caa1935c47b1902d/pkg/util/admission/granter.go#L623)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you rename this to `TestingDisableSkipEnforcement` and also the same name change in the `GrantCoordinator` field.
</blockquote></details>

:+1:
___
*[pkg/util/admission/granter.go, line 639 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mx0Oq_F-o-ZM6rwwN94:-Mx1Ajjb4hPQIURTQdWO:bmeyrxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbc99d7be71af1326498e44caa1935c47b1902d/pkg/util/admission/granter.go#L639)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you change this `SQLSQLResponseBurstTokens` comment to the same TODO as above. It is not unused.
</blockquote></details>

:+1:  
___
*[pkg/util/admission/granter.go, line 645 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mx0PDieCyaJgc45jPHr:-Mx1CWni9wU7urCubHeX:bpp8rfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbc99d7be71af1326498e44caa1935c47b1902d/pkg/util/admission/granter.go#L645)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Is differing from Go defaults how we normally handle such overrides? What if one wanted to override something to 0?
</blockquote></details>

Good question, I'm not aware of any conventions for this.
___
*[pkg/sql/run_control_test.go, line 754 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-Mx0Lk4X4qpdZZmnfRmp:-Mx1Ah3qA7ZDFoCdXHh9:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cbc99d7be71af1326498e44caa1935c47b1902d/pkg/sql/run_control_test.go#L754)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I'd also add that we do assert that the actual table ID matches this value below, after the server is started.
</blockquote></details>

:+1:

**Review by sumeerbhola — APPROVED** [review]
:lgtm: modulo kv signoff on the error handling change

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76736#-:-MxA6pjtBRc3DB0ZNGuO:b-ucpxlh)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @cucaroach, @msirek, @rytaft, @sumeerbhola, and @yuzefovich)

**Pull request merged**
